### PR TITLE
Goyal1092 multihierarchy

### PR DESCRIPTION
**Pull request by jbothma**
## Checklist

- [ ]  🚀 is the code ready to be merged and go live?
- [ ]  🛠 does it work (build) locally

### Pull Request

- [ ]  📰 good title
- [ ]  📝good description
- [ ]  🔖 issue linked
- [ ]  📖 changelog filled out

### Commits

- [ ]  commits are clean
- [ ]  commit messages are clean

### Code Quality

- [ ]  🚧 no commented out code
- [ ]  🖨 no unnecessary logging
- [ ]  🎱 no magic numbers
- [ ]  black was run locally (as part of the pre-commit hook)

### Testing

- [ ]  ✅ added (appropriate) unit tests
- [ ]  💢 edge cases in tests were considered
- [ ]  ✅ ran tests locally & are passing

### CODE REVIEW

File: wazimap_ng/datasets/views.py [hunk 1]
```diff
@@ -176,16 +176,19 @@ class GeographyHierarchyViewset(viewsets.ReadOnlyModelViewSet):
...
-    max-results number of results to be returned [default is 30] 
+    max-results number of results to be returned [default is 30]
     """
     profile = get_object_or_404(Profile, pk=profile_id)
-    version = profile.geography_hierarchy.root_geography.version
-    
+    version_name = request.GET.get("version", None)
+    if not version:
+        version_name = profile.geography_hierarchy.configuration.get("default_version", None)
+    version = get_object_or_404(models.Version, name=version_name)
```
Comment (jbothma) [inline wazimap_ng/datasets/views.py:190]: make sure we're only returning results relevant to the profile

File: wazimap_ng/general/views.py [hunk 2]
```diff
@@ -24,12 +24,17 @@
...
-    profile_js = profile_serializers.ExtendedProfileSerializer(profile, geography)
+    versions_list = profile.geography_hierarchy.configuration.get("versions", [])
+    if not version_name:
+        version_name = profile.geography_hierarchy.configuration.get("default_version", None)
+    if not version_name or version_name not in versions_list:
+        raise Http404
+
+    version = get_object_or_404(dataset_models.Version, name=version_name)
+    geography = dataset_models.Geography.objects.get(code=geography_code, versions=version)
+    profile_js = profile_serializers.ExtendedProfileSerializer(profile, geography, version)
```
Comment (jbothma) [inline wazimap_ng/general/views.py:37]: make sure we're getting a geography relevant to the profile

File: wazimap_ng/general/views.py [hunk 3]
```diff
@@ -52,17 +57,43 @@ def consolidated_profile_helper(profile_id, geography_code):
...
+
+    version = request.GET.get('version', None)
+    versions_list = profile.geography_hierarchy.configuration.get("versions", [])
+    if not version:
+        version = profile.geography_hierarchy.configuration.get("default_version", None)
+    if not version or version not in versions_list:
+        raise Http40
+    version = get_object_or_404(dataset_models.Version, name=version)
+
+    geography = dataset_models.Geography.objects.get(code=geography_code, versions=version)
```
Comment (jbothma) [inline wazimap_ng/general/views.py:78]: make sure we're selecting a boundary relevant to the profile

File: wazimap_ng/general/views.py [hunk 4]
```diff
@@ -52,17 +57,43 @@ def consolidated_profile_helper(profile_id, geography_code):
...
-    geography = dataset_models.Geography.objects.get(code=geography_code, version=version)
-    return Response(point_views.boundary_point_count_helper(profile, geography))
+
+    version = request.GET.get('version', None)
+    versions_list = profile.geography_hierarchy.configuration.get("versions", [])
+    if not version:
+        version = profile.geography_hierarchy.configuration.get("default_version", None)
+    if not version or version not in versions_list:
+        raise Http40
+    version = get_object_or_404(dataset_models.Version, name=version)
```
Comment (jbothma) [inline wazimap_ng/general/views.py:76]: version name and version object is being assigned to the same variable
the variable's type changes. This is bad.

File: wazimap_ng/profile/views.py [hunk 5]
```diff
@@ -62,10 +62,14 @@ def retrieve(self, request, *args, **kwargs):
...
     profile = get_object_or_404(models.Profile, pk=profile_id)
-    version = profile.geography_hierarchy.root_geography.version
-    geography = get_object_or_404(Geography, code=geography_code, version=version)
+    version_name = request.GET.get("version", None)
+    if not version_name:
+        version_name = profile.geography_hierarchy.configuration.get("default_version", None)
 
-    js = serializers.ExtendedProfileSerializer(profile, geography)
+    version = get_object_or_404(Version, name=version_name)
+    geography = get_object_or_404(Geography, code=geography_code, versions=version)
```
Comment (jbothma) [inline wazimap_ng/profile/views.py:70]: might wanna make sure we're getting a geo related to the profile

File: wazimap_ng/datasets/models/geography.py [hunk 6]
```diff
@@ -30,10 +38,10 @@ class Geography(MP_Node, BaseModel):
     name = models.CharField(max_length=50)
     code = models.CharField(max_length=20)
     level = models.CharField(max_length=20)
-    version = models.CharField(max_length=20, blank=True)
+    versions = models.ManyToManyField(Version, blank=True)
```
Comment (jbothma) [inline wazimap_ng/datasets/models/geography.py:41]: So Geography 1 ----- * GeographyBoundary
And Geography * ----- * Version
And GeographyBoundary * ------ * Version

What's ensuring that Geographies relate to exactly the versions their boundaries relate to?

Do we even use Geography.versions anywhere?

File: wazimap_ng/datasets/models/geography.py [hunk 7]
```diff
@@ -65,23 +72,39 @@ def get_child_boundaries(self):
...
         return child_types
 
+
+    def get_child_geographies(self, version):
+        child_geographies = self.get_children().filter(geographyboundary__version=version)
+        return child_geographies
+
+    @property
+    def versions(self):
+        return []
```
Comment (goyal1092) [inline wazimap_ng/datasets/models/geography.py:91]: Remove this property for now

File: wazimap_ng/datasets/views.py [hunk 8]
```diff
@@ -226,9 +229,9 @@ def geography_ancestors(request, geography_code, version):
     Returns parent geographies of the given geography code
     Return a 404 HTTP response if the is the code is not found
     """
-    geos = models.Geography.objects.filter(code=geography_code, version=version)
+    geos = models.Geography.objects.filter(code=geography_code, versions__name=version)
```
Comment (goyal1092) [inline wazimap_ng/datasets/views.py:232]: versions removed from geography model

File: wazimap_ng/points/services/locations.py [hunk 9]
```diff
@@ -2,21 +2,23 @@
...
     if category is not None:
         queryset = queryset.filter(category=category)
 
     if geography_code is not None:
-        version = profile.geography_hierarchy.root_geography.version
-        geography = Geography.objects.get(code=geography_code, version=version) 
-        boundary = GeographyBoundary.objects.get(geography=geography)
+        version = Version.objects.get(name=version_name)
+        boundary = GeographyBoundary.objects.get(geography__code=geography_code, version=version)
+        geography = boundary.geography
```
Comment (goyal1092) [inline wazimap_ng/points/services/locations.py:21]: Unused geography variable

File: wazimap_ng/points/services/locations.py [hunk 10]
```diff
@@ -2,21 +2,23 @@
...
 
     if category is not None:
         queryset = queryset.filter(category=category)
 
     if geography_code is not None:
-        version = profile.geography_hierarchy.root_geography.version
-        geography = Geography.objects.get(code=geography_code, version=version) 
-        boundary = GeographyBoundary.objects.get(geography=geography)
+        version = Version.objects.get(name=version_name)
+        boundary = GeographyBoundary.objects.get(geography__code=geography_code, version=version)
```
Comment (goyal1092) [inline wazimap_ng/points/services/locations.py:20]: Might throw error if:
* Geo code does not exists. We are not checking if geo exist in views
* will throw error if boundary does not exist for version and geo code

File: wazimap_ng/boundaries/management/commands/loadshp.py [hunk 11]
```diff
@@ -33,28 +35,64 @@ def check_shapefile(self, shapefile_path):
...
-        g = Geography.add_root(**fields)
-        GeographyBoundary.objects.create(geography=g, geom=geo_shape, area=area)
 
-    def process_node(self, fields, geo_shape):
+        try:
+            g = Geography.objects.get(code=fields["code"], geographyhierarchy__name=hierarchy_name)
+        except Geography.DoesNotExist:
+            g = Geography.add_root(**fields)
+            self.ensure_hierachy_exists(hierarchy_name, create_hierarchy, g)
+        # g.versions.add(version)
```
Comment (jbothma) [inline wazimap_ng/boundaries/management/commands/loadshp.py:58]: commented code

File: wazimap_ng/boundaries/management/commands/loadshp.py [hunk 12]
```diff
@@ -33,28 +35,64 @@ def check_shapefile(self, shapefile_path):
...
 
         try:
-            geography = Geography.objects.get(code__iexact=fields["code"], version=version)
+            geography = parent_geography.get_descendants().get(code__iexact=fields["code"])
         except Geography.DoesNotExist:
-            geography = parent_geography.add_child(code=fields["code"], level=fields["level"], name=fields["name"], version=version)
-            GeographyBoundary.objects.create(geography=geography, geom=geo_shape, area=area)
+            geography = parent_geography.add_child(**fields)
+
+        # geography.versions.add(version)
```
Comment (jbothma) [inline wazimap_ng/boundaries/management/commands/loadshp.py:89]: commented unneeded code

File: wazimap_ng/datasets/admin/dataset_admin.py [hunk 13]
```diff
@@ -56,22 +56,18 @@ class DatasetAdmin(DatasetBaseAdminModel):
...
             "fields": (
-                "profile", "name", "geography_hierarchy",
-                "permission_type"
+                "profile", "name", "permission_type"
             )
         }),
         ("Dataset Imports", {
             "fields": (
-                "content_type", "import_dataset", "imported_dataset",
+                "version", "content_type", "import_dataset", "imported_dataset",
```
Comment (jbothma) [inline wazimap_ng/datasets/admin/dataset_admin.py:64]: we probably wanna restrict versions to those relevant to the selected profile

File: wazimap_ng/datasets/admin/forms/dataset_admin_form.py [hunk 14]
```diff
@@ -34,9 +33,4 @@ def __init__(self, *args, **kwargs):
                 ).first()
 
             if filtered_profile:
-                self.fields["profile"].initial = filtered_profile
-
-                for hierarchy in hierarchies:
-                    hp = hierarchy.profile_set.all()
-                    if hp.count() == 1 and hp.first() == filtered_profile:
-                        self.fields["geography_hierarchy"].initial = hierarchy
```
Comment (jbothma) [inline wazimap_ng/datasets/admin/forms/dataset_admin_form.py:42]: looks like this is how a relevant initial hierarchy used to be selected. Perhaps this is what we want for version too?
